suite/tests/client-interface/drbbdup-drwrap-test.appdll.c

```diff
@@ -56,7 +48,33 @@ run_tests(void)
 #else
     __asm__ __volatile__("nop; nop; nop; nop");
 #endif
+}
+
+int EXPORT
+wrapme(int x)
+{
+    print("%s: arg %d\n", __FUNCTION__, x);
+    if (x % 2 == 0) {
+        /* Switch in the middle of a wrapped function. */
+        switch_modes();
+    }
+    return x;
+}
+
+void
+run_tests(void)
+{
+    /* First, a regular pre-and-post-wrapped instance. */
+    print("first wrapme returned %d\n", wrapme(1));
+    /* Now we'll have a pre but not post b/c we'll swap in the middle. */
     print("second wrapme returned %d\n", wrapme(2));
+    /* Now we swap back and will have no pre or post. */
```

Comment:
So when we swap first, we only have a pre, but not post. When we swap back, we don't have any; I'm wondering if a user might expect a post, because the instrumentation changed during the wrapped-func.

I'm also wondering if there can be cases where the user is keeping some state between the pre and post callbacks, and whether it's better to always do both or none. It may be difficult to do that though.

Reply:
It is not feasible to do none since you wouldn't know in the pre whether the phase is going to change.

For a user who wants the post to happen regardless of the phase, they would just not use this inverted control mode.  (Not 100% sure everything works though with drwrap using regular drmgr compare to the user using drbbdup but I think it does: drwrap will inserts its calls *before* the drbbdup dispatch so they'll happen for all cases.  I don't think it's worth our time right now to go make sure that works; wait for someone who needs it.)